File: JS/easy.js

/**
 * 
 * @name Make-Array-Elements-Equal-To-Zero
 * 
 * @description Start by selecting a starting position curr such that nums[curr] == 0, and choose a movement direction of either left or right.

After that, you repeat the following process:

If curr is out of the range [0, n - 1], this process ends.
If nums[curr] == 0, move in the current direction by incrementing curr if you are moving right, or decrementing curr if you are moving left.
Else if nums[curr] > 0:
Decrement nums[curr] by 1.
Reverse your movement direction (left becomes right and vice versa).
Take a step in your new direction.
A selection of the initial position curr and movement direction is considered valid if every element in nums becomes 0 by the end of the process.

Note: Please do not copy the description during the contest to maintain the integrity of your submissions.©leetcode
 */
var countValidSelections = function(nums) {
    function canMakeAllZeros(start, direction) {
        let temp = [...nums];
        let n = temp.length;
        let curr = start;

        while (curr >= 0 && curr < n) {
          if (temp[curr] === 0) {
            curr += direction;
          } else {
            temp[curr] -= 1;
            direction *= -1;
            curr += direction;
          }
        }
        return temp.every(num => num === 0);
      }

      let validSelections = 0;
    
      for (let i = 0; i < nums.length; i++) {
        if (nums[i] === 0) {
          if (canMakeAllZeros(i, -1)) {
            validSelections++;
          }
          if (canMakeAllZeros(i, 1)) {
            validSelections++;
          }
        }
      }

  return validSelections;
};

/**
 * 
 * @name 
 * 
 * @description 
 * 
 * 
 */
var canAliceWin = function(n) {
    let arem = 10;
    let brem = 9;
    let remain = n;
    let isAliceTurn = true;
    while (remain > 0) {
        if (isAliceTurn) {
            remain -= arem;
            brem = arem - 1;
            isAliceTurn = false
            if(remain < 0) return false
            if(remain === 0) return true
        } else {
            remain -= brem;
            arem = brem - 1
            isAliceTurn = true
            if(remain < 0) return true
            if(remain === 0) return false
        }
    }
}


/**
 * 
 * @name permutation
 * 
 * @description 
 * 
 * 
 */
var permute = function(nums) {
  const res = [];
  function backTrack(start){
      if(start === nums.length){
          res.push([...nums]);
          return;
      }
      for(let i=start; i<nums.length; i++){
          [nums[start], nums[i]] = [nums[i], nums[start]];
          backTrack(start+1);
          [nums[start], nums[i]] = [nums[i], nums[start]];
      }
  }
  backTrack(0);
  return res;
};


/**
* 
* @name Diameter-Of-Binary-Tree
* @alias BST DFS
* @description 
* 
* 
*/
var diameterOfBinaryTree = function(root) {
  let maxLen = 0;
  const dfs = (root) => {
      if(!root){
          return 0;
      }
      let right = dfs(root.right);
      let left = dfs(root.left);
      maxLen = Math.max(maxLen, right+left);
      return 1+Math.max(right, left);
  }
  dfs(root);
  return maxLen;
};


/**
* 
* @name Reverse-LinkedList
* @alias List
* @description 
* 
* 
*/
var reverseList = function(head) {
  if(!head){
      return head;
  }
  let prev = null, current = head;
  while(current){
      const next = current.next;
      current.next = prev;
      prev = current;
      current = next;
  }
  return prev;
};


/**
 * 
 * @name Has-Cycle
 * 
 * @description Linked List Cycle
 * 
 * 
 */
function hasCycle(head) {
  if(!head || !head.next){
      return false;
  }
  let slow = head;
  let fast = head.next;
  while(slow !== fast){
      if(!fast || !fast.next){
          return false
      }
      slow = slow.next;
      fast = fast.next.next;
  }
  return true;
};

/**
 * 
 * @name Majority-Element
 * 
 * @description Elements That Occured Most In An Array
 * 
 * 
 */
//My Solution!
var majorityElement = function(nums) {
  const yt = nums.length/2;
  let hash = {};
  for(let i=0; i<=nums.length-1; i++){
      if(nums[i] in hash){
          hash[nums[i]] += 1;
      }else {
          hash[nums[i]] = 1;
      }
  }
  let maxEle = -Infinity;
  Object.entries(hash).map(([key, val]) => {
      if(val > yt) {
          maxEle = Math.max(maxEle, Number(key));
      }
  })
  return maxEle;
};
//Optimal Solution -- Boyer Moore Voting Algorithm
var majorityElement = function(nums) {
  let count = 0;
  let candidate = null;
  for(let i=0; i<=nums.length-1; i++){
      if(count===0){
          candidate = nums[i];
      }

      if(nums[i]===candidate){
          count +=1
      }else {
          count -=1
      }
  }
  return candidate;
};

/**
 * 
 * @name Invert-Tree
 * 
 * @description Invert-Binary-Tree
 * 
 * 
 */
var invertTree = function(root) {
  if(!root){
      return root;
  }
  let left = invertTree(root.left);
  let right = invertTree(root.right);
  root.left = right;
  root.right = left;
  return root;
};

//max parenthensis depth
var maxDepth = function(s) {
  let max = 0;
  let count = [];
  for(let i=0; i<s.length; i++){
      if(s[i]==="("){
          count.push("(");
      }
      if(s[i]===")"){
          max = Math.max(max, count.length);
          count.pop("(");
      }
  }
  return max;
};

File: GO/easy.go
package golang

import (
	"fmt"
	"math"
)

/**
 *
 * @name
 *
 * @description
 *
 *
 */
func canAliceWin(n int) bool {
	arem := 10
	brem := 9
	remain := n
	isAliceTurn := true

	var final bool

	for remain > 0 {
		if isAliceTurn {
			remain -= arem
			brem = arem - 1
			isAliceTurn = false
			if remain < 0 {
				final = false
				break
			}
			if remain == 0 {
				final = true
				break
			}
		} else {
			remain -= brem
			arem = brem - 1
			isAliceTurn = true
			if remain < 0 {
				final = true
				break
			}
			if remain == 0 {
				final = false
				break
			}
		}
	}
	return final
}

/**
 *
 * @name Has-Cycle
 *
 * @description Linked List Cycle
 *
 *
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}
	slow := head
	fast := head.Next
	for slow != fast {
		if fast == nil || fast.Next == nil {
			return false
		}
		slow = slow.Next
		fast = fast.Next.Next
	}
	return true
}

/**
 *
 * @name Majority-Element
 *
 * @description Elements That Occured Most In An Array
 *
 * Optimal Solution -- Boyer Moore Voting Algorithm
 */
func majorityElement(nums []int) int {
	count := 0
	candidate := 0
	for i := 0; i <= len(nums)-1; i++ {
		if count == 0 {
			candidate = nums[i]
		}
		if candidate == nums[i] {
			count++
		} else {
			count--
		}
	}
	return candidate
}

/**
 *
 * @name Invert-Tree
 *
 * @description Invert-Binary-Tree
 *
 *
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func invertTree(root *TreeNode) *TreeNode {
	if root == nil {
		return root
	}
	left := invertTree(root.Left)
	right := invertTree(root.Right)
	root.Left = right
	root.Right = left
	return root
}

// BFS Graph Implementation
func bfs(start string, graph map[string][]string) {
	queue := []string{start}
	visted := make(map[string]bool)
	for len(queue) > 0 {
		// fmt.Println("Q deQ", queue)
		current := queue[0] //get index 1 of the queue
		queue = queue[1:]   //remove first index from queue
		if visted[current] {
			continue
		}
		visted[current] = true
		fmt.Println("Current City: ", current)
		for _, neigh := range graph[current] {
			if !visted[neigh] {
				queue = append(queue, neigh)
			}
		}
	}
}
func bfsMain() {
	graph := map[string][]string{
		"Abuja":      {"Minna", "Lokoja", "Jos", "Kaduna"},
		"Minna":      {"Abuja", "Kaduna", "Ilorin"},
		"Lokoja":     {"Abuja", "Benin City", "Makurdi"},
		"Jos":        {"Abuja", "Bauchi", "Kaduna", "Gombe"},
		"Kaduna":     {"Abuja", "Kano", "Zaria", "Jos"},
		"Ilorin":     {"Minna", "Ibadan", "Oshogbo"},
		"Benin City": {"Lokoja", "Asaba", "Warri"},
		"Makurdi":    {"Lokoja", "Enugu", "Jalingo"},
		"Bauchi":     {"Jos", "Gombe", "Maiduguri"},
		"Gombe":      {"Jos", "Bauchi", "Yola"},
		"Kano":       {"Kaduna", "Katsina", "Dutse"},
		"Zaria":      {"Kaduna", "Kano"},
		"Asaba":      {"Benin City", "Onitsha", "Awka"},
		"Onitsha":    {"Asaba", "Awka", "Enugu"},
		"Enugu":      {"Onitsha", "Awka", "Abakaliki"},
		"Ibadan":     {"Ilorin", "Oshogbo", "Abeokuta", "Lagos"},
		"Oshogbo":    {"Ibadan", "Ado-Ekiti", "Akure"},
		"Ado-Ekiti":  {"Oshogbo", "Akure"},
		"Akure":      {"Ado-Ekiti", "Benin City"},
		"Abeokuta":   {"Ibadan", "Lagos"},
		"Lagos":      {"Abeokuta", "Ibadan"},
	}
	fmt.Println("Traversal starts, Cities in Nigeria by distance from Abuja FCT") //Change the city name to find nearest to city you input
	bfs("Abuja", graph)
}

// Edge represents a weighted edge in the graph
type Edge struct {
	node   string
	weight int
}

// Dijkstra's algorithm implementation
func dijkstra(start string, graph map[string][]Edge) map[string]int {
	distances := make(map[string]int)
	for node := range graph {
		distances[node] = math.MaxInt32 // Initialize distances to infinity
	}
	distances[start] = 0 // Distance to the start node is 0

	// Simulate a priority queue using a slice
	queue := []string{start}

	return distances
}
